src/cargo/ops/cargo_compile/packages.rs

```diff
@@ -0,0 +1,218 @@
+use std::collections::BTreeSet;
+
+use crate::core::Package;
+use crate::core::{PackageIdSpec, Workspace};
+use crate::util::restricted_names::is_glob_pattern;
+use crate::util::CargoResult;
+
+use anyhow::{bail, Context as _};
+
+/// Represents the selected pacakges that will be built.
+///
+/// Generally, it represents the combination of all `-p` flag. When working within
+/// a workspace, `--exclude` and `--workspace` flags also contribute to it.
+#[derive(PartialEq, Eq, Debug)]
+pub enum Packages {
```

Comment:
Side note: long term, I wonder where this should belong as this seems like a more fundamental CLI package selection struct that is useful outside of compilation.

For example, `clap-cargo` tries to imitate this for general use.  `cargo-release` is at least one program that uses that logic.  `cargo-upgrade` had another implementation of it until we switched to oly operating on the workspace.

Being buried in `cargo_compile` makes this less discoverable.

Reply:
Fully agree. That's why even I hate the lost of git history so much but still did this 🫠